articles/node-url-shortener/index.md

```diff
@@ -0,0 +1,231 @@
+const express = require('express')
+
+const router = express.Router()
+
+const Url = require('../models/Url')
+
+// : app.get(/:code)
+
+// @route       GET /:code
+// @description    Redirect to the long/original URL 
+router.get('/:code', async(req, res)=>{
+    try{
+        const url = await Url.findOne({urlCode: req.params.code})
+        if(url){
+            return res.redirect(url.longUrl)
+        }
+        else{
+            return res.status(404).json('No URL Found')
+        }
+
+    }
+    catch(err){
+        console.error(err)
+        res.status(500).json('Server Error')
+    }
+})
+
+
+module.exports = router
+```
+
+Let us now understand the above code. To set up a router in express, we need to import the `express` module and initialize the `express.Router()` method. This API route requires the database model we created to save the short URL that will be generated. For this to happen, we need to import `valid-url` for validation and `short-id` that will create the unique ID for our short URL. Our localhost application will use a base URL (`baseUrl` variable) as `http:\\localhost:5000`. We now need to extract the `longUrl` from our request body, validate id with the `valid-url` package by executing the method `validUrl.isUri(baseUrl)` and pass the URL as the argument. If the URL is valid, we generate a short code using the `const urlCode = shortid.generate()` and append it to our base URL before saving it to the database. This process is asynchronous and that is why we use promises or async-await syntax.
+
+### The final `index.js` Entry file
+
+```js
+const express = require("express")
+const app= express()
+
+// Database config
+const connection =  require('./config/db')
+connection.once('open', ()=>console.log('DB Connected'))
+connection.on('error', ()=>console.log('Error'))
+
+// Routes Config
+app.use(express.json({extended: false})) //parse incoming request body in JSON format.
+app.use('/', require('./routes/redirect'))
+app.use('/api/url', require('./routes/url'))
+
+//Listen for incoming requests
+const PORT = process.env.PORT || 5000
+app.listen(PORT, console.log(`server started, listening PORT ${PORT}`))
+```
+
+To make our routes work, we have to use a [middleware pattern](https://expressjs.com/en/guide/using-middleware.html#middleware.application) is express. Middleware functions have access to request (req) and response object (res) in the application’s request-response cycle. This explains the `app.use()` methods. The first middleware allows our application to parse incoming request data format in JSON format. The `app.use('/', require('./routes/redirect'))` is the has the base URI that will configure the redirect route. In our POST route, the base URL is `/api/url` and the middleware as `app.use('/api/url', require('./routes/url'))`. Next, we need to test the application in postman.
+
+### A Working Demo In Postman
+
+![postman_test_gif](/engineering-education/node-url-shortener/postman.gif)
+
+### Conclusion
```

Comment:
It would be helpful if you could upload the final code on GitHub and paste a link to the repository